samples/algorithms/sudoku-grover/ColoringGroverWithConstraints.qs

```diff
@@ -0,0 +1,395 @@
+    /// the same color
+    /// ## startingColorConstraints
+    /// The array of (Vertex#,Color) specifying the dissallowed colors for vertices
+    ///
+    /// # Output
+    /// A unitary operation that applies `oracle` on the target register if the control 
+    /// register state corresponds to the bit mask `bits`.
+    ///
+    /// # Example
+    /// Consider the following 4x4 Sudoku puzzle 
+    ///     -----------------
+    ///     |   |   | 2 | 3 |
+    ///     -----------------
+    ///     |   |   | 0 | 1 |
+    ///     -----------------
+    ///     | 1 | 2 | 3 | 0 |
+    ///     -----------------
+    ///     | 3 | 0 | 1 | 2 |
+    ///     -----------------
+    ///  The challenge is to fill the empty squares with numbers 0 to 3
+    ///  that are unique in row, column and the top left 2x2 square
+    ///  This is a graph coloring problem where the colors are 0 to 3
+    ///  and the empty cells are the vertices. The vertices can be defined as   
+    ///     -----------------
+    ///     | 0 | 1 |   |   |
+    ///     -----------------
+    ///     | 2 | 3 |   |   |
+    ///     -----------------
+    ///     |   |   |   |   |
+    ///     -----------------
+    ///     |   |   |   |   |
+    ///     -----------------
+    /// The graph is
+    ///  0---1
+    ///  | X |
+    ///  1---2
+    /// i.e. every vertex is connected to each other
+    /// But we also require that 
+    ///    - vertices 0 and 1 do not get colors 2 and 3
+    ///    - vertices 2 and 3 do not get colors 3 and 0
+    ///    - vertices 0 and 2 do not get colors 1 and 3
+    ///    - vertices 1 and 3 do not get colors 2 and 0 
+    /// This results in edges (vertices that can not be same color)
+    /// edges = [(1, 0),(2, 0),(3, 0),(3, 1),(3, 2)]
+    /// This is saying that vertex 1 can not have same color as vertex 0 etc.
+    /// and startingColorConstraints = [(0, 1),(0, 3),(0, 2),(1, 2),(1, 0),
+    ///                    (1, 3),(2, 1),(2, 3),(2, 0),(3, 2),(3, 0),(3, 1)]
+    /// This is saying that vertex 0 is not allowed to have values 1,3,2
+    /// and vertex 1 is not allowed to have values 2,0,3
+    /// and vertex 2 is not allowed to have values 1,3,0
+    /// and vertex 3 is not allowed to have values 2,0,1
+    /// A valid graph coloring solution is: [0,1,2,3]
+    /// i.e. vextex 0 has color 0, vertex 1 has color 1 etc.
+    operation VertexColoringOracle (numVertices : Int, bitsPerColor : Int, edges : (Int, Int)[],  
+        startingColorConstraints : (Int, Int)[], 
+        colorsRegister : Qubit[], 
+        target : Qubit) : Unit is Adj+Ctl {
+        let nEdges = Length(edges);
+        let nStartingColorConstraints = Length(startingColorConstraints);
+        // we are looking for a solution that
+        // (a) has no edge with same color at both ends and 
+        // (b) has no Vertex with a color that violates the starting color constraints
+        using (conflictQubits = Qubit[nEdges + nStartingColorConstraints]) {
```

Comment:
It may make the code more readable here to allocate two arrays, one for edge conflicts the other for initial condition conflicts, then append them at line 152 when controlling `X`.